ext/lmdb-safe/lmdb-safe.hh

```diff
@@ -12,9 +15,95 @@
 #include <mutex>
 #include <vector>
 #include <algorithm>
+#include <bit>
+#include "../../pdns/misc.hh"
 
 using std::string_view;
 
+#ifndef DNSDIST
+
+// FIXME do something more portable than __builtin_bswap64
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+#define _LMDB_SAFE_BSWAP64MAYBE(x) __builtin_bswap64(x)
+#else
+#define _LMDB_SAFE_BSWAP64MAYBE(x) (x)
+#endif
+
+namespace {
+  class __attribute__((__packed__)) LSheader {
+  public:
+    uint64_t d_timestamp;
+    uint64_t d_txnid;
+    uint8_t d_version;
+    uint8_t d_flags;
+    uint32_t d_reserved;
+    uint16_t d_numextra;
+
+    LSheader(uint64_t timestamp, uint64_t txnid, uint8_t flags=0, uint8_t version=0, uint8_t numextra=0):
+      d_timestamp(_LMDB_SAFE_BSWAP64MAYBE(timestamp)),
+      d_txnid(_LMDB_SAFE_BSWAP64MAYBE(txnid)),
+      d_version(version),
+      d_flags(flags),
+      d_reserved(0),
+      d_numextra(numextra)
+    {
+
+    }
+
+    std::string toString() {
+      return std::string((char*)this, sizeof(*this));
+    }
+
+
+  };
+
+  static_assert(sizeof(LSheader)==24, "LSheader size is wrong");
+
+  const size_t LS_HEADER_SIZE=sizeof(LSheader);  // FIXME: rename this so all code that relies on it breaks - it needs to use LScheckHeaderAndGetSize below
+  const size_t LS_BLOCK_SIZE=8;
+  const uint8_t LS_FLAG_DELETED = 0x01;
+
+// maybe put these functions in a namespace with a name?
+
+  // this also returns a pointer to the string's data. Do not hold on to it too long!
+  LSheader* LSassertFixedHeaderSize(std::string &val) {
+    // cerr<<"val.size()="<<val.size()<<endl;
+    if (val.size() < LS_HEADER_SIZE) {
+      throw std::runtime_error("LSheader too short");
+    }
+
+    return (LSheader*)val.data();
```

Comment:
Since the data typically points somewhere into an LMDB page, there is no guarantee that it is 8-byte aligned.

It may be faster to manually assemble all the individual bytes on access, than to use a clever bswap64 only to incur unaligned load penalties on access.

Do other parts of the LMDB backend code assume that the application value is aligned?

Reply:
> It may be faster to manually assemble all the individual bytes on access, than to use a clever bswap64 only to incur unaligned load penalties on access.

probably true. This was simpler for now :)

> Do other parts of the LMDB backend code assume that the application value is aligned?

I don't think so, but I'm not 100% sure.

Reply:
I guess the only fields that are currently accessed are the byte `flags` and `version` fields, or which case it does not matter, and the `numExtra` field for which it does.

Reply:
Agreed, should not be hard, but let's do it in alpha2.

Reply:
did it here anyway